### DISCUSSION TITLE
Best way to get tracking data via Python on another machine?

### DISCUSSION THREAD

**Discussion by Tagirijus** [almarklein/timetagger#507] [Q&A] [answered]
Hey there,
I have a server and I am running a Timetagger instance. It works perfectly and I am able to track my time there.

Now I wonder what would be the best way to get time tracking data via Python locally on my other machine. Especially when it comes to searching for specific time tracking data with a certain set of tags. Is this somehow possible?

I looked at timetagger_cli, but this seems to be a CLI application only. Even the Python module just prints out what you can do via terminal, but does not return e.g. Python dict or lists or so.

Looking forward to replies! Thanks.

**Comment by almarklein** [comment, marked as answer]
You'd basically have to use the web api: https://timetagger.readthedocs.io/en/stable/webapi/  You can peak at timetagger_cli for inspiration.

The server is actually rather stupid and does not support any kind of search functionality (yet). The timetagger web app has all your records, and "simply" sync any changes that the app makes to them.

>  Is this somehow possible?

So ... not really, at the moment. The best you can get is download all your records, preferably cache them on disk, and search to these locally with some basic Python string searches.

**Reply by Tagirijus** [reply]
Thanks for your reply. I already started this way. It just feels a bit odd how I did it so far: I try to fetch all records and then do the query stuff locally (without caching so far).

And there already just came an issue up: when I would like to fetch records before 1970-01-01, the timestamp would become negative. This, on the other hand, seem to raise an error for the GET /records timerange parameter, which will result in something like `timerange=-2208992400-1733914824`. Am i doing something wrong here, or is fetching records with a timestamp < 0 just impossible?

To be fair: I am not even that old that I personally could have tracked time before timestamp 0 ... still I wanted to not exclude such cases, just in case I might make my script public.

Edit:
I get `bad request: /records timerange needs 2 numbers (timestamps)` as an error / output with the above mentioned parameter. Snippet of the code with which I am doing this:

```python
headers = {
    'authtoken': api_token
}

r = requests.get(
    'https://timetagger.mydomain.com/timetagger/api/v2/records',
    headers=headers,
    params={
        'timerange': '-2208992400-1733914824'
    }
)

obj = r.json()
return obj['records']
```

**Reply by almarklein** [reply]
> or is fetching records with a timestamp < 0 just impossible?

Exactly. There cannot exit records older than 1970 :), so using 0 will catch all records.

**Reply by Tagirijus** [reply]
Thanks for your answer. Unlucky time travellers then. (;
